zuul-core/src/main/java/com/netflix/zuul/netty/server/psk/TlsPskHandler.java

```diff
@@ -0,0 +1,499 @@
+        super();
+        this.registry = registry;
+        this.externalTlsPskProvider = externalTlsPskProvider;
+    }
+
+    @Override
+    public void bind(ChannelHandlerContext ctx, SocketAddress localAddress, ChannelPromise promise) throws Exception {
+        ctx.bind(localAddress, promise);
+    }
+
+    @Override
+    public void connect(
+            ChannelHandlerContext ctx, SocketAddress remoteAddress, SocketAddress localAddress, ChannelPromise promise)
+            throws Exception {
+        ctx.connect(remoteAddress, localAddress, promise);
+    }
+
+    @Override
+    public void disconnect(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ctx.disconnect(promise);
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ctx.close(promise);
+    }
+
+    @Override
+    public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        ctx.deregister(promise);
+    }
+
+    @Override
+    public void read(ChannelHandlerContext ctx) throws Exception {
+        ctx.read();
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (!(msg instanceof ByteBuf byteBufMsg)) {
+            ctx.write(msg, promise);
+            return;
+        }
+        byte[] appDataBytes = byteBufMsg.hasArray() ? byteBufMsg.array() : readDirect(byteBufMsg);
+        ReferenceCountUtil.safeRelease(byteBufMsg);
+        tlsPskServerProtocol.writeApplicationData(appDataBytes, 0, appDataBytes.length);
+        int availableOutputBytes = tlsPskServerProtocol.getAvailableOutputBytes();
+        if (availableOutputBytes != 0) {
+            byte[] outputBytes = new byte[availableOutputBytes];
+            tlsPskServerProtocol.readOutput(outputBytes, 0, availableOutputBytes);
+            ctx.write(Unpooled.wrappedBuffer(outputBytes), promise)
+                    .addListener(ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE);
+        }
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        ctx.flush();
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+        final byte[] bytesRead = in.hasArray() ? in.array() : readDirect(in);
+        try {
+            tlsPskServerProtocol.offerInput(bytesRead);
+        } catch (TlsFatalAlert tlsFatalAlert) {
+            writeOutputIfAvailable(ctx);
+            return;
+        }
+        writeOutputIfAvailable(ctx);
+        final int appDataAvailable = tlsPskServerProtocol.getAvailableInputBytes();
+        if (appDataAvailable > 0) {
+            byte[] appData = new byte[appDataAvailable];
+            tlsPskServerProtocol.readInput(appData, 0, appDataAvailable);
+            out.add(Unpooled.wrappedBuffer(appData));
```

Comment:
Do we need to pass this data down to other handlers?

Reply:
Based on my understanding, the fact that TlsPskHandler is a `ByteToMessageDecoder`, adding the decoded output to `List<Object> out` does the job of passing down the data to other handlers. LMK if I'm missing the point that you are making.